File: rs/minsk/src/analysis/syntax/tree.rs
use crate::analysis::diagnostic::Diagnostic;
use crate::analysis::text::source_text::SourceText;

use super::lexer::Lexer;
use super::node::unit::CompilationUnitSyntax;
use super::parser::Parser;
use super::token::SyntaxToken;

pub struct SyntaxTree {
    pub root: CompilationUnitSyntax,
    pub diagnostics: Vec<Diagnostic>,
    pub source_text: SourceText,
}

impl SyntaxTree {
    fn new(source_text: SourceText) -> Self {
        let mut parser = Parser::new(source_text);
        let root = parser.parse_compilation_unit();
        let (source_text, diagnostics) = parser.unpack();
        Self {
            root,
            diagnostics,
            source_text,
        }
    }

    pub fn parse(text: &str) -> Self {
        Self::new(SourceText::from(text.chars().collect::<Vec<_>>()))
    }

    pub fn parse_tokens(text: &str) -> Vec<SyntaxToken> {
        Lexer::new(&SourceText::from(text.chars().collect::<Vec<_>>())).into_only_tokens()
    }
}

File: py/minsk/analysis/syntax/tree.py
from minsk.analysis.syntax.expression import ExpressionSyntax
from minsk.analysis.syntax.token import SyntaxToken


class SyntaxTree:
    root: ExpressionSyntax
    end_of_file_token: SyntaxToken
    diagnostics: tuple[str, ...]

    def __init__(
        self,
        root: ExpressionSyntax,
        end_of_file_token: SyntaxToken,
        diagnostics: tuple[str, ...],
    ):
        self.root = root
        self.end_of_file_token = end_of_file_token
        self.diagnostics = diagnostics
